lib/ukcpio/cpio.c

```diff
@@ -0,0 +1,312 @@
+		else if (buf[i] >= 'A' && buf[i] <= 'F')
+			val += (buf[i] - 'A') + 10;
+		else if (buf[i] >= 'a' && buf[i] <= 'f')
+			val += (buf[i] - 'a') + 10;
+		else
+			return 0;
+	}
+	return val;
+}
+
+/**
+ * Create absolute path with prefix.
+ *
+ * @param prefix
+ *  Prefix to the path.
+ * @param path
+ *  Path to file or directory.
+ * @return
+ *  Returns the absolute path with prefix.
+ */
+static char *
+absolute_path(const char *prefix, const char *path)
+{
+	int add_slash;
+	size_t prefix_len;
+	size_t path_len;
+	size_t abs_path_len;
+	char *abs_path;
+
+	UK_ASSERT(prefix);
+	UK_ASSERT(path);
+
+	prefix_len = strlen(prefix);
+	path_len = strlen(path);
+
+	add_slash = prefix[prefix_len - 1] == '/' ? 0 : 1;
+	abs_path_len = prefix_len + add_slash + path_len + 1;
+
+	abs_path = malloc(abs_path_len);
+
+	if (abs_path == NULL)
+		return NULL;
+
+	memcpy(abs_path, prefix, prefix_len);
+	if (add_slash)
+		abs_path[prefix_len] = '/';
+	memcpy(&abs_path[prefix_len + add_slash], path, path_len);
+
+	abs_path[abs_path_len - 1] = '\0';
+	return abs_path;
+}
+
+/**
+ * Reads the section to the dest from a given a CPIO header.
+ *
+ * @param header_ptr
+ *  Pointer to the CPIO header.
+ * @param dest
+ *  Destination path to extract the current header.
+ * @param last
+ *  The size of the CPIO section.
+ * @return
+ *  Returns 0 on success or one of ukcpio_error enum.
+ */
+static enum ukcpio_error
+read_section(struct cpio_header **header_ptr,
+				const char *dest, uintptr_t last)
+{
+	enum ukcpio_error error = UKCPIO_SUCCESS;
+	int fd;
+	struct cpio_header *header;
+	char *path_from_root;
+	mode_t header_mode;
+	uint32_t header_filesize;
+	uint32_t header_namesize;
+	char *data_location;
+	uint32_t bytes_to_write;
+	int bytes_written;
+	struct cpio_header *next_header;
+
+	if (strcmp(filename(*header_ptr), "TRAILER!!!") == 0) {
+		*header_ptr = NULL;
+		return UKCPIO_SUCCESS;
+	}
+
+	if (!valid_magic(*header_ptr)) {
+		*header_ptr = NULL;
+		return -UKCPIO_INVALID_HEADER;
+	}
+
+	UK_ASSERT(dest);
+
+	header = *header_ptr;
+	path_from_root = absolute_path(dest, filename(header));
+
+	if (path_from_root == NULL) {
+		*header_ptr = NULL;
+		error = -UKCPIO_NOMEM;
+		goto out;
+	}
+
+	header_mode = GET_MODE(header);
+	header_filesize = S8HEX_TO_U32(header->filesize);
+	header_namesize = S8HEX_TO_U32(header->namesize);
+
+	if ((uintptr_t)header + sizeof(struct cpio_header) > last) {
+		*header_ptr = NULL;
+		error = -UKCPIO_MALFORMED_FILE;
+		goto out;
+	}
+
+	if (IS_FILE(header_mode) && header_filesize != 0) {
+		uk_pr_info("Extracting %s...\n", path_from_root);
+		fd = open(path_from_root, O_CREAT | O_RDWR);
+
+		if (fd < 0) {
+			*header_ptr = NULL;
+			error = -UKCPIO_FILE_CREATE_FAILED;
+			goto out;
+		}
+
+		data_location = (char *)ALIGN_4(
+				(char *)(header) + sizeof(struct cpio_header)
+				+ header_namesize);
+
+		if ((uintptr_t)data_location + header_filesize > last) {
+			*header_ptr = NULL;
+			error = -UKCPIO_MALFORMED_FILE;
```

Comment:
`uk_pr_err`?

Reply:
```diff
+                       uk_pr_err("Could not read CPIO file!\n");
```
?

Reply:
Hum...
```c
uk_pr_err("Malformed CPIO input");
```
?

I would also change the error code and remove `FILE` since it is not necessarily a file, just a mapped buffer